staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go

```diff
@@ -73,7 +75,17 @@ func withWaitGroup(handler http.Handler, longRunning apirequest.LongRunningReque
 			return
 		}
 
+		if terminator != nil && longRunningRequest {
+			newCtx, cancelFn := terminator.Attach(ctx)
+			ctx = newCtx
+			defer cancelFn()
+		}
+
 		defer wg.Done()
-		handler.ServeHTTP(w, req)
+		handler.ServeHTTP(w, req.WithContext(ctx))
 	})
 }
+
+type RequestTerminator interface {
+	Attach(context.Context) (context.Context, context.CancelFunc)
+}
```

Comment:
I would also claim that having `StartTerminating` method as part of this interface would be more intuitive - it would better show what it is supposed to do...